src/pictureNamingTask.ts

```diff
@@ -0,0 +1,236 @@
+
+  // if all images have been shown clear the set
+  if (indiciesSelected.size === array.length) {
+    indiciesSelected.clear();
+  }
+
+  do {
+    const [newRandomIndex, newRng] = prand.uniformIntDistribution(
+      0,
+      array.length - 1,
+      rng,
+    );
+    rng = newRng;
+    randomIndex = newRandomIndex;
+
+    if (!indiciesSelected.has(randomIndex)) {
+      indiciesSelected.add(randomIndex);
+      foundUnique = true;
+    }
+  } while (!foundUnique);
+
+  console.log("Selected Index:", randomIndex);
+  console.log("Selected Indices Set:", indiciesSelected);
+
+  let result = [array[randomIndex]];
+  console.log("***********************");
+  console.log("Result from random index:");
+  console.table(result);
+  console.log("***********************");
+  return result;
+}
+
+// draw an image at random from the bank depending on the difficulty_level selected
+function createStimuli(
+  difficultyLevel: number,
+  language: string,
+  clearSet: boolean,
+) {
+  if (clearSet === true) {
+    indiciesSelected.clear();
+  }
+  try {
+    let imgList = imageDB;
+    imgList = imgList.filter(
+      (image: any) =>
+        Number(image.difficultyLevel) === difficultyLevel &&
+        image.language === language,
+    );
+    console.log("filtered");
+    console.log(imgList);
+    let result = getRandomElementWithSeed(imgList);
+    return result;
+  } catch (error) {
+    console.error("Error fetching and parsing data:", error);
+  }
+}
+
+//****************************
+//********EXPERIMENT**********
+//****************************
+// a timeline is a set of trials
+// a trial is a single object eg htmlKeyboardResponse etc ...
+const timeline: any[] = [];
+export default function pictureNamingTask(difficultyLevelParam: number) {
+  let experiment_stimuli = createStimuli(difficultyLevelParam, language, false);
+  let currentDifficultyLevel: number = difficultyLevelParam;
+  if (difficultyLevelParam) {
+    const jsPsych = initJsPsych({
+      on_finish: function () {
+        transformAndDownload(jsPsych.data);
+      },
+    });
+
+    const welcome = {
+      type: HtmlKeyboardResponsePlugin,
+      stimulus: `${welcomeText}`,
+    };
+
+    const preload = {
+      type: PreloadPlugin,
+      auto_preload: true,
+      show_progress_bar: true,
+      message: `<p>loading stimulus</p>`,
+    };
+
+    const blankPage = {
+      type: HtmlKeyboardResponsePlugin,
+      stimulus: "",
+    };
+    const showImg = {
+      type: ImageKeyboardResponsePlugin,
+      stimulus_height: 600,
+      stimulus: jsPsych.timelineVariable("stimulus"),
+    };
+
+    const logging = {
+      type: SurveyHtmlFormPlugin,
+      preamble: function () {
+        const html = `<h3>Correct response: </h3>
+                    <p>${jsPsych.evaluateTimelineVariable("correctResponse")}</p>
+                    <img src="${jsPsych.evaluateTimelineVariable("stimulus")}" width="300" height="300">`;
+
+        return html;
+      },
+      html: `
```

Comment:
look into importing html as a string using the `?raw` suffix in Vite. It is not your fault, but this looks gross.


See https://vitejs.dev/guide/assets#importing-asset-as-string

Reply:
Highly error prone and insecure

Reply:
https://owasp.org/www-community/attacks/xss/

Reply:
not an issue with current implementation. this plugin doesn't run <script> and it's all done on client side.